README.md
PHP Nano Http Status
============

[![Latest Stable Version](https://poser.pugx.org/gino-pane/nano-http-status/v/stable)](https://packagist.org/packages/gino-pane/nano-http-status)
[![Build Status](https://travis-ci.org/GinoPane/php-nano-rest.svg?branch=master)](https://travis-ci.org/GinoPane/php-nano-rest)
[![Maintainability](https://img.shields.io/codeclimate/maintainability/GinoPane/php-nano-http-status.svg)](https://codeclimate.com/github/GinoPane/php-nano-http-status/maintainability)
[![Test Coverage](https://img.shields.io/codeclimate/coverage/github/GinoPane/php-nano-http-status.svg)](https://codeclimate.com/github/GinoPane/php-nano-http-status/test_coverage)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/GinoPane/php-nano-http-status/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/GinoPane/php-nano-http-status/?branch=master)
[![License](https://poser.pugx.org/gino-pane/nano-http-status/license)](https://packagist.org/packages/gino-pane/nano-http-status)
[![composer.lock](https://poser.pugx.org/gino-pane/nano-http-status/composerlock)](https://packagist.org/packages/gino-pane/nano-http-status)
[![Total Downloads](https://poser.pugx.org/gino-pane/nano-http-status/downloads)](https://packagist.org/packages/gino-pane/nano-http-status)

Truly minimalistic and self-contained package to handle HTTP statuses and relevant description messages. Such packages as 
[Teapot](https://github.com/shrikeh/teapot) or [Httpstatus](https://github.com/lukasoppermann/http-status) are nice and famous, but still
either too heavy and over-complicated or simply does not fit to Gino Pane's high code quality standards.

Requirements
============

* PHP >= 7.0;

Features
========

* The full list of HTTP status codes as readable constants;
* relevant status messages;
* ability to customize messages;
* ability to easily detect the class of the status;
* no dependencies in production;
* integrated tools for code quality, testing and building docs.

Installation
============

    composer require gino-pane/nano-http-status

Basic Usage
===========

Check existence of the code (using numeric codes or nice readable constants):

    (new NanoHttpStatus())->statusExists(200); //true
    (new NanoHttpStatus())->statusExists(400); //true
    (new NanoHttpStatus())->statusExists(451); //true
    (new NanoHttpStatus())->statusExists(511); //true
    (new NanoHttpStatus())->statusExists(522); //false
    
Detect the class of the code using code numbers:

    (new NanoHttpStatus())->isInformational(NanoHttpStatus::HTTP_OK); //false
    (new NanoHttpStatus())->isSuccess(202);         //true
    (new NanoHttpStatus())->isRedirection(301);     //true
    (new NanoHttpStatus())->isClientError(404);     //true
    (new NanoHttpStatus())->isServerError(NanoHttpStatus::HTTP_BAD_REQUEST); //false
    
Get status message by status code:

    (new NanoHttpStatus())->getMessage(200); //OK
    (new NanoHttpStatus())->getMessage(451); //Unavailable For Legal Reasons
    (new NanoHttpStatus())->getMessage(452); //Undefined Status
    
Set localization mapping and get custom status messages:

    $status = new NanoHttpStatus([
        NanoHttpStatus::HTTP_BAD_REQUEST => 'Very bad request',
        NanoHttpStatus::HTTP_BAD_GATEWAY => 'Not so bad gateway'
    ]);
    
    $status->getMessage(400); //'Very bad request'
    $status->getMessage(502); //'Not so bad gateway'
    
Please note, that ```NanoHttpStatus``` itself does not throw any exceptions for invalid statuses.

Useful Tools
============

Running Tests:
--------

    php vendor/bin/phpunit
 
 or 
 
    composer test

Code Sniffer Tool:
------------------

    php vendor/bin/phpcs --standard=PSR2 src/
 
 or
 
    composer psr2check

Code Auto-fixer:
----------------

    php vendor/bin/phpcbf --standard=PSR2 src/ 
    
 or
 
    composer psr2autofix
 
 
Building Docs:
--------

    php vendor/bin/phpdoc -d "src" -t "docs"
 
 or 
 
    composer docs

Changelog
=========

To keep track, please refer to [CHANGELOG.md](https://github.com/GinoPane/php-nano-http-status/blob/master/CHANGELOG.md).

Contributing
============

1. Fork it.
2. Create your feature branch (git checkout -b my-new-feature).
3. Make your changes.
4. Run the tests, adding new ones for your own code if necessary (phpunit).
5. Commit your changes (git commit -am 'Added some feature').
6. Push to the branch (git push origin my-new-feature).
7. Create new pull request.

Also please refer to [CONTRIBUTION.md](https://github.com/GinoPane/php-nano-http-status/blob/master/CONTRIBUTION.md).

License
=======

Please refer to [LICENSE](https://github.com/GinoPane/php-nano-http-status/blob/master/LICENSE).
 
Notes
=====
 
Powered by [composer-package-template](https://github.com/GinoPane/composer-package-template)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

Types of changes

* **Features** for initial package features.
* **Added** for new features.
* **Changed** for changes in existing functionality.
* **Deprecated** for soon-to-be removed features.
* **Removed** for now removed features.
* **Fixed** for any bug fixes.
* **Security** in case of vulnerabilities.

## [Unreleased]

## 1.0.0 - 2017-12-12

### Features

* The full list of HTTP status codes as readable constants;
* relevant status messages;
* ability to customize messages;
* ability to easily detect the class of the status;
* no dependencies in production;
* integrated tools for code quality, testing and building docs.

## 0.1.0 - 2017-12-11

### Features

* The initial release after clean installation from composer-package-template

[Unreleased]: https://github.com/GinoPane/nano-http-status/compare/v0.1.0...HEAD


tests/suites/main/YourClassTest.php
<?php 

namespace GinoPane\NanoHttpStatus;

use PHPUnit\Framework\TestCase;

/**
*  Corresponding class to test YourClass class
*
*  For each class in your library, there should be a corresponding unit test
*
*  @author yourname
*/
class YourClassTest extends TestCase
{
    /**
     * Just check if the YourClass has no syntax errors
     */
    public function testIsThereAnySyntaxError()
    {
        $object = new YourClass();

        $this->assertTrue(is_object($object));
    }

    /**
     * Test the only existing method of the class
     *
     * @dataProvider getNamesAndGreetings
     *
     * @param $name
     * @param $expected
     */
    public function testSayHello($name, $expected)
    {
        $object = new YourClass();

        $this->assertEquals($expected, $object->sayHello($name));
    }

    /**
     * Data for sayHello
     *
     * @return array
     */
    public function getNamesAndGreetings(): array
    {
        return [
            ['world', "Hello World!"],
            ['World', "Hello World!"]
        ];
    }
}

tests/suites/main/NanoHttpStatusTest.php
<?php

namespace GinoPane\NanoHttpStatus;

use Generator;
use ReflectionClass;
use PHPUnit\Framework\TestCase;

/**
 * Class NanoHttpStatusTest
 *
 * @author Sergey <Gino Pane> Karavay
 */
class NanoHttpStatusTest extends TestCase
{
    /**
     * Just check if the YourClass has no syntax errors
     */
    public function testIsThereAnySyntaxError()
    {
        $object = new NanoHttpStatus();

        $this->assertTrue($object instanceof NanoHttpStatus);
    }

    /**
     * @dataProvider getAllValidStatusConstants
     *
     * @param int $code
     */
    public function testIfAllStatusCodesExist(int $code)
    {
        $this->assertTrue((new NanoHttpStatus())->statusExists($code));
    }

    /**
     * @dataProvider getInformationalCodes
     *
     * @param int $code
     */
    public function testIfInformationalCodesAreDetected(int $code)
    {
        $this->assertTrue((new NanoHttpStatus())->isInformational($code));
    }

    /**
     * @dataProvider getSomeInvalidStatusConstants
     *
     * @param int $code
     */
    public function testIfNonInformationalCodesAreDetectedWithNonExistent(int $code)
    {
        $this->assertFalse((new NanoHttpStatus())->isInformational($code));
    }

    /**
     * @dataProvider getClientErrorCodes
     *
     * @param int $code
     */
    public function testIfNonInformationalCodesAreDetectedWithAnotherClass(int $code)
    {
        $this->assertFalse((new NanoHttpStatus())->isInformational($code));
    }

    /**
     * @dataProvider getSuccessCodes
     *
     * @param int $code
     */
    public function testIfSuccessCodesAreDetected(int $code)
    {
        $this->assertTrue((new NanoHttpStatus())->isSuccess($code));
    }

    /**
     * @dataProvider getSomeInvalidStatusConstants
     *
     * @param int $code
     */
    public function testIfNonSuccessCodesAreDetectedWithNonExistent(int $code)
    {
        $this->assertFalse((new NanoHttpStatus())->isSuccess($code));
    }

    /**
     * @dataProvider getClientErrorCodes
     *
     * @param int $code
     */
    public function testIfNonSuccessCodesAreDetectedWithAnotherClass(int $code)
    {
        $this->assertFalse((new NanoHttpStatus())->isSuccess($code));
    }

    /**
     * @dataProvider getRedirectionCodes
     *
     * @param int $code
     */
    public function testIfRedirectionCodesAreDetected(int $code)
    {
        $this->assertTrue((new NanoHttpStatus())->isRedirection($code));
    }

    /**
     * @dataProvider getSomeInvalidStatusConstants
     *
     * @param int $code
     */
    public function testIfNonRedirectionCodesAreDetectedWithNonExistent(int $code)
    {
        $this->assertFalse((new NanoHttpStatus())->isRedirection($code));
    }

    /**
     * @dataProvider getClientErrorCodes
     *
     * @param int $code
     */
    public function testIfNonRedirectionCodesAreDetectedWithAnotherClass(int $code)
    {
        $this->assertFalse((new NanoHttpStatus())->isRedirection($code));
    }

    /**
     * @dataProvider getClientErrorCodes
     *
     * @param int $code
     */
    public function testIfClientErrorCodesAreDetected(int $code)
    {
        $this->assertTrue((new NanoHttpStatus())->isClientError($code));
    }

    /**
     * @dataProvider getSomeInvalidStatusConstants
     *
     * @param int $code
     */
    public function testIfNonClientErrorCodesAreDetectedWithNonExistent(int $code)
    {
        $this->assertFalse((new NanoHttpStatus())->isClientError($code));
    }

    /**
     * @dataProvider getServerErrorCodes
     *
     * @param int $code
     */
    public function testIfNonClientErrorCodesAreDetectedWithAnotherClass(int $code)
    {
        $this->assertFalse((new NanoHttpStatus())->isClientError($code));
    }

    /**
     * @dataProvider getServerErrorCodes
     *
     * @param int $code
     */
    public function testIfServerErrorCodesAreDetected(int $code)
    {
        $this->assertTrue((new NanoHttpStatus())->isServerError($code));
    }

    /**
     * @dataProvider getSomeInvalidStatusConstants
     *
     * @param int $code
     */
    public function testIfNonServerErrorCodesAreDetectedWithNonExistent(int $code)
    {
        $this->assertFalse((new NanoHttpStatus())->isServerError($code));
    }

    /**
     * @dataProvider getClientErrorCodes
     *
     * @param int $code
     */
    public function testIfNonServerErrorCodesAreDetectedWithAnotherClass(int $code)
    {
        $this->assertFalse((new NanoHttpStatus())->isServerError($code));
    }

    /**
     * @dataProvider getSomeInvalidStatusConstants
     *
     * @param int $code
     */
    public function testIfSomeStatusCodesDontExist(int $code)
    {
        $this->assertFalse((new NanoHttpStatus())->statusExists($code));
    }

    /**
     * @dataProvider getDefaultStatusMessages
     *
     * @param int $code
     * @param string $expectedMessage
     */
    public function testIfStatusMessagesAreFetchedCorrectly(int $code, string $expectedMessage)
    {
        $this->assertEquals($expectedMessage, (new NanoHttpStatus())->getMessage($code));
    }

    /**
     * @dataProvider getSomeInvalidStatusConstants
     *
     * @param int $code
     */
    public function testIfUndefinedStatusIsReturnedForInvalidCode(int $code)
    {
        $this->assertEquals('Undefined Status', (new NanoHttpStatus())->getMessage($code));
    }

    public function testIfLocalizationArrayReplacesMessages()
    {
        $status = new NanoHttpStatus([
            400 => 'Very bad request',
            502 => 'Not so bad gateway'
        ]);

        $this->assertEquals('Very bad request', $status->getMessage(400));
        $this->assertEquals('Not so bad gateway', $status->getMessage(502));
    }

    /**
     * @return Generator
     */
    public function getAllValidStatusConstants()
    {
        $class = new ReflectionClass(NanoHttpStatusInterface::class);

        foreach ($class->getConstants() as $constant => $value) {
            if (stripos($constant, 'http_') === 0) {
                yield [$value];
            }
        }
    }

    /**
     * @return array
     */
    public function getSomeInvalidStatusConstants()
    {
        return [
            [0],
            [1],
            [131],
            [225],
            [331],
            [445],
            [522],
            [601]
        ];
    }

    /**
     * @return array
     */
    public function getInformationalCodes()
    {
        return [
            [100],
            [101],
            [102],
            [103],
        ];
    }

    /**
     * @return array
     */
    public function getSuccessCodes()
    {
        return [
            [200],
            [201],
            [202],
            [203],
            [204],
            [205],
            [206],
            [207],
            [208],
            [226]
        ];
    }

    /**
     * @return array
     */
    public function getRedirectionCodes()
    {
        return [
            [300],
            [301],
            [302],
            [303],
            [304],
            [305],
            [307],
            [308],
        ];
    }

    /**
     * @return array
     */
    public function getClientErrorCodes()
    {
        return [
            [400],
            [401],
            [402],
            [403],
            [404],
            [405],
            [407],
            [408],
            [451]
        ];
    }

    /**
     * @return array
     */
    public function getServerErrorCodes()
    {
        return [
            [500],
            [501],
            [502],
            [507],
            [510]
        ];
    }

    /**
     * @return Generator
     */
    public function getDefaultStatusMessages()
    {
        $class = new ReflectionClass(NanoHttpStatus::class);
        $statuses = $class->getProperty('defaultStatuses');
        $statuses->setAccessible(true);
        $defaultMessages = $statuses->getValue();

        foreach ($defaultMessages as $code => $message) {
            yield [$code, $message];
        }
    }
}


src/NanoHttpStatus.php
<?php

namespace GinoPane\NanoHttpStatus;

/**
 * Class NanoHttpStatus
 *
 * @author Sergey <Gino Pane> Karavay
 */
class NanoHttpStatus implements NanoHttpStatusInterface
{
    /**
     * Default placeholder for non-existent status
     */
    const UNDEFINED_STATUS_MESSAGE = 'Undefined Status';

    /**
     * HTTP status common descriptions
     *
     * @var array
     */
    private static $defaultStatuses = [
        self::HTTP_CONTINUE => 'Continue',
        self::HTTP_SWITCHING_PROTOCOLS => 'Switching Protocols',
        self::HTTP_PROCESSING => 'Processing',
        self::HTTP_EARLY_HINTS => 'Early Hints',
        self::HTTP_OK => 'OK',
        self::HTTP_CREATED => 'Created',
        self::HTTP_ACCEPTED => 'Accepted',
        self::HTTP_NON_AUTHORITATIVE_INFORMATION => 'Non-Authoritative Information',
        self::HTTP_NO_CONTENT => 'No Content',
        self::HTTP_RESET_CONTENT => 'Reset Content',
        self::HTTP_PARTIAL_CONTENT => 'Partial Content',
        self::HTTP_MULTI_STATUS => 'Multi-Status',
        self::HTTP_ALREADY_REPORTED => 'Already Reported',
        self::HTTP_IM_USED => 'IM Used',
        self::HTTP_MULTIPLE_CHOICES => 'Multiple Choices',
        self::HTTP_MOVED_PERMANENTLY => 'Moved Permanently',
        self::HTTP_FOUND => 'Found',
        self::HTTP_SEE_OTHER => 'See Other',
        self::HTTP_NOT_MODIFIED => 'Not Modified',
        self::HTTP_USE_PROXY => 'Use Proxy',
        self::HTTP_TEMPORARY_REDIRECT => 'Temporary Redirect',
        self::HTTP_PERMANENTLY_REDIRECT => 'Permanent Redirect',
        self::HTTP_BAD_REQUEST => 'Bad Request',
        self::HTTP_UNAUTHORIZED => 'Unauthorized',
        self::HTTP_PAYMENT_REQUIRED => 'Payment Required',
        self::HTTP_FORBIDDEN => 'Forbidden',
        self::HTTP_NOT_FOUND => 'Not Found',
        self::HTTP_METHOD_NOT_ALLOWED => 'Method Not Allowed',
        self::HTTP_NOT_ACCEPTABLE => 'Not Acceptable',
        self::HTTP_PROXY_AUTHENTICATION_REQUIRED => 'Proxy Authentication Required',
        self::HTTP_REQUEST_TIMEOUT => 'Request Timeout',
        self::HTTP_CONFLICT => 'Conflict',
        self::HTTP_GONE => 'Gone',
        self::HTTP_LENGTH_REQUIRED => 'Length Required',
        self::HTTP_PRECONDITION_FAILED => 'Precondition Failed',
        self::HTTP_PAYLOAD_TOO_LARGE => 'Payload Too Large',
        self::HTTP_URI_TOO_LONG => 'URI Too Long',
        self::HTTP_UNSUPPORTED_MEDIA_TYPE => 'Unsupported Media Type',
        self::HTTP_RANGE_NOT_SATISFIABLE => 'Requested Range Not Satisfiable',
        self::HTTP_EXPECTATION_FAILED => 'Expectation Failed',
        self::HTTP_I_AM_A_TEAPOT => 'I\'m a teapot',
        self::HTTP_MISDIRECTED_REQUEST => 'Misdirected Request',
        self::HTTP_UNPROCESSABLE_ENTITY => 'Unprocessable Entity',
        self::HTTP_LOCKED => 'Locked',
        self::HTTP_FAILED_DEPENDENCY => 'Failed Dependency',
        self::HTTP_UPGRADE_REQUIRED => 'Upgrade Required',
        self::HTTP_PRECONDITION_REQUIRED => 'Precondition Required',
        self::HTTP_TOO_MANY_REQUESTS => 'Too Many Requests',
        self::HTTP_REQUEST_HEADER_FIELDS_TOO_LARGE => 'Request Header Fields Too Large',
        self::HTTP_UNAVAILABLE_FOR_LEGAL_REASONS => 'Unavailable For Legal Reasons',
        self::HTTP_INTERNAL_SERVER_ERROR => 'Internal Server Error',
        self::HTTP_NOT_IMPLEMENTED => 'Not Implemented',
        self::HTTP_BAD_GATEWAY => 'Bad Gateway',
        self::HTTP_SERVICE_UNAVAILABLE => 'Service Unavailable',
        self::HTTP_GATEWAY_TIMEOUT => 'Gateway Timeout',
        self::HTTP_VERSION_NOT_SUPPORTED => 'HTTP Version Not Supported',
        self::HTTP_VARIANT_ALSO_NEGOTIATES => 'Variant Also Negotiates',
        self::HTTP_INSUFFICIENT_STORAGE => 'Insufficient Storage',
        self::HTTP_LOOP_DETECTED => 'Loop Detected',
        self::HTTP_NOT_EXTENDED => 'Not Extended',
        self::HTTP_NETWORK_AUTHENTICATION_REQUIRED => 'Network Authentication Required',
    ];

    /**
     * Array of ready-to-use status messages
     *
     * @var string[]
     */
    private $statuses = [];

    public function __construct(array $localize = [])
    {
        $this->statuses = self::$defaultStatuses;

        if (!empty($localize)) {
            $this->statuses = array_intersect_key($localize, $this->statuses) + $this->statuses;
        }
    }

    /**
     * Returns 'true' if status code exists, 'false' otherwise
     *
     * @param int $code
     *
     * @return bool
     */
    public function statusExists(int $code): bool
    {
        return isset($this->statuses[$code]);
    }

    /**
     * Returns relevant status message if status exists, 'Undefined Status' is returned otherwise
     *
     * @param int $code
     *
     * @return string
     */
    public function getMessage(int $code): string
    {
        return $this->statusExists($code) ? $this->statuses[$code] : self::UNDEFINED_STATUS_MESSAGE;
    }

    /**
     * Returns 'true' if status exists and belongs to "informational" (1xx) statuses, 'false' otherwise
     *
     * @param int $code
     *
     * @return bool
     */
    public function isInformational(int $code): bool
    {
        return $this->statusExists($code) && ($this->extractFirstDigit($code) == self::CLASS_INFORMATIONAL);
    }

    /**
     * Returns 'true' if status exists and belongs to "success" (2xx) statuses, 'false' otherwise
     *
     * @param int $code
     *
     * @return bool
     */
    public function isSuccess(int $code): bool
    {
        return $this->statusExists($code) && ($this->extractFirstDigit($code) == self::CLASS_SUCCESS);
    }

    /**
     * Returns 'true' if status exists and belongs to "redirection" (3xx) statuses, 'false' otherwise
     *
     * @param int $code
     *
     * @return bool
     */
    public function isRedirection(int $code): bool
    {
        return $this->statusExists($code) && ($this->extractFirstDigit($code) == self::CLASS_REDIRECTION);
    }

    /**
     * Returns 'true' if status exists and belongs to "client error" (4xx) statuses, 'false' otherwise
     *
     * @param int $code
     *
     * @return bool
     */
    public function isClientError(int $code): bool
    {
        return $this->statusExists($code) && ($this->extractFirstDigit($code) == self::CLASS_CLIENT_ERROR);
    }

    /**
     * Returns 'true' if status exists and belongs to "server error" (5xx) statuses, 'false' otherwise
     *
     * @param int $code
     *
     * @return bool
     */
    public function isServerError(int $code): bool
    {
        return $this->statusExists($code) && ($this->extractFirstDigit($code) == self::CLASS_SERVER_ERROR);
    }

    /**
     * Extracts the first digit of the number
     *
     * @param int $code
     *
     * @return int
     */
    private function extractFirstDigit(int $code): int
    {
        return (int)substr((string)$code, 0, 1);
    }
}


src/NanoHttpStatusInterface.php
<?php

namespace GinoPane\NanoHttpStatus;

/**
 * Interface NanoHttpStatusInterface
 *
 * @author Sergey <Gino Pane> Karavay
 */
interface NanoHttpStatusInterface
{
    /**
     * The list of available HTTP status codes
     *
     * All codes that do not have explicitly defined RFC fall into RFC7231
     *
     * @link http://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml
     */
    const HTTP_CONTINUE = 100;
    const HTTP_SWITCHING_PROTOCOLS = 101;
    const HTTP_PROCESSING = 102;                        // RFC2518
    const HTTP_EARLY_HINTS = 103;                       // RFC-ietf-httpbis-early-hints-05
    const HTTP_OK = 200;
    const HTTP_CREATED = 201;
    const HTTP_ACCEPTED = 202;
    const HTTP_NON_AUTHORITATIVE_INFORMATION = 203;
    const HTTP_NO_CONTENT = 204;
    const HTTP_RESET_CONTENT = 205;
    const HTTP_PARTIAL_CONTENT = 206;                   // RFC7233
    const HTTP_MULTI_STATUS = 207;                      // RFC4918
    const HTTP_ALREADY_REPORTED = 208;                  // RFC5842
    const HTTP_IM_USED = 226;                           // RFC3229
    const HTTP_MULTIPLE_CHOICES = 300;
    const HTTP_MOVED_PERMANENTLY = 301;
    const HTTP_FOUND = 302;
    const HTTP_SEE_OTHER = 303;
    const HTTP_NOT_MODIFIED = 304;                      // RFC7232
    const HTTP_USE_PROXY = 305;
    const HTTP_TEMPORARY_REDIRECT = 307;
    const HTTP_PERMANENTLY_REDIRECT = 308;              // RFC7238
    const HTTP_BAD_REQUEST = 400;
    const HTTP_UNAUTHORIZED = 401;                      // RFC7235
    const HTTP_PAYMENT_REQUIRED = 402;
    const HTTP_FORBIDDEN = 403;
    const HTTP_NOT_FOUND = 404;
    const HTTP_METHOD_NOT_ALLOWED = 405;
    const HTTP_NOT_ACCEPTABLE = 406;
    const HTTP_PROXY_AUTHENTICATION_REQUIRED = 407;     //RFC7235
    const HTTP_REQUEST_TIMEOUT = 408;
    const HTTP_CONFLICT = 409;
    const HTTP_GONE = 410;
    const HTTP_LENGTH_REQUIRED = 411;
    const HTTP_PRECONDITION_FAILED = 412;               //RFC7232
    const HTTP_PAYLOAD_TOO_LARGE = 413;
    const HTTP_URI_TOO_LONG = 414;
    const HTTP_UNSUPPORTED_MEDIA_TYPE = 415;
    const HTTP_RANGE_NOT_SATISFIABLE = 416;             //RFC7233
    const HTTP_EXPECTATION_FAILED = 417;
    const HTTP_I_AM_A_TEAPOT = 418;                     // RFC2324
    const HTTP_MISDIRECTED_REQUEST = 421;               // RFC7540
    const HTTP_UNPROCESSABLE_ENTITY = 422;              // RFC4918
    const HTTP_LOCKED = 423;                            // RFC4918
    const HTTP_FAILED_DEPENDENCY = 424;                 // RFC4918
    const HTTP_UPGRADE_REQUIRED = 426;                  // RFC2817
    const HTTP_PRECONDITION_REQUIRED = 428;             // RFC6585
    const HTTP_TOO_MANY_REQUESTS = 429;                 // RFC6585
    const HTTP_REQUEST_HEADER_FIELDS_TOO_LARGE = 431;   // RFC6585
    const HTTP_UNAVAILABLE_FOR_LEGAL_REASONS = 451;     // RFC7725
    const HTTP_INTERNAL_SERVER_ERROR = 500;
    const HTTP_NOT_IMPLEMENTED = 501;
    const HTTP_BAD_GATEWAY = 502;
    const HTTP_SERVICE_UNAVAILABLE = 503;
    const HTTP_GATEWAY_TIMEOUT = 504;
    const HTTP_VERSION_NOT_SUPPORTED = 505;
    const HTTP_VARIANT_ALSO_NEGOTIATES = 506;           // RFC2295
    const HTTP_INSUFFICIENT_STORAGE = 507;              // RFC4918
    const HTTP_LOOP_DETECTED = 508;                     // RFC5842
    const HTTP_NOT_EXTENDED = 510;                      // RFC2774
    const HTTP_NETWORK_AUTHENTICATION_REQUIRED = 511;   // RFC6585

    /**
     * The list of possible classes of status codes
     */
    const CLASS_INFORMATIONAL   = 1; //1xx
    const CLASS_SUCCESS         = 2; //2xx
    const CLASS_REDIRECTION     = 3; //3xx
    const CLASS_CLIENT_ERROR    = 4; //4xx
    const CLASS_SERVER_ERROR    = 5; //5xx
}
